substrate/pkg/controller/substrate/infrastructure/natgateway.go

```diff
@@ -0,0 +1,127 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package infrastructure
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/awslabs/kubernetes-iteration-toolkit/substrate/pkg/apis/v1alpha1"
+	"github.com/awslabs/kubernetes-iteration-toolkit/substrate/pkg/utils/discovery"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type NatGateway struct {
+	EC2 *ec2.EC2
+}
+
+func (n *NatGateway) Create(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
+	if substrate.Status.Infrastructure.VPCID == nil ||
+		substrate.Status.Infrastructure.PrivateRouteTableID == nil ||
+		len(substrate.Status.Infrastructure.PublicSubnetIDs) == 0 ||
+		substrate.Status.Infrastructure.ElasticIpIDForNatGW == nil {
+		return reconcile.Result{Requeue: true}, nil
+	}
+	natGW, err := n.ensure(ctx, substrate)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if _, err := n.EC2.CreateRouteWithContext(ctx, &ec2.CreateRouteInput{
+		RouteTableId:         substrate.Status.Infrastructure.PrivateRouteTableID,
```

Comment:
does this route table contains all the private subnets cidrs ? Just want to make sure every private subnet has a route to NATGW.

Reply:
Yes all the subnets share the same NAT GW and all of them are added to the route table.